src/objective-c/ProtoRPC/ProtoRPC.m

```diff
@@ -23,9 +23,166 @@
+                       responseHandler:(id<GRPCResponseHandler>)handler
+                           callOptions:(GRPCCallOptions *)callOptions
+                         responseClass:(Class)responseClass {
+  if ((self = [super init])) {
+    _call = [[GRPCStreamingProtoCall alloc] initWithRequestOptions:requestOptions
+                                                   responseHandler:handler
+                                                       callOptions:callOptions
+                                                     responseClass:responseClass];
+    [_call writeWithMessage:message];
+    [_call finish];
+  }
+  return self;
+}
+
+- (void)cancel {
+  [_call cancel];
+  _call = nil;
+}
+
+@end
+
+@interface GRPCStreamingProtoCall ()<GRPCResponseHandler>
+
+@end
+
+@implementation GRPCStreamingProtoCall {
+  GRPCRequestOptions *_requestOptions;
+  id<GRPCResponseHandler> _handler;
+  GRPCCallOptions *_callOptions;
+  Class _responseClass;
+
+  GRPCCall2 *_call;
+  dispatch_queue_t _dispatchQueue;
+}
+
+- (instancetype)initWithRequestOptions:(GRPCRequestOptions *)requestOptions
+                       responseHandler:(id<GRPCResponseHandler>)handler
+                           callOptions:(GRPCCallOptions *)callOptions
+                         responseClass:(Class)responseClass {
+  if ((self = [super init])) {
+    _requestOptions = [requestOptions copy];
+    _handler = handler;
+    _callOptions = [callOptions copy];
+    _responseClass = responseClass;
+    _dispatchQueue = dispatch_queue_create(nil, DISPATCH_QUEUE_SERIAL);
+
+    [self start];
+  }
+  return self;
+}
+
+- (void)start {
+  _call = [[GRPCCall2 alloc] initWithRequestOptions:_requestOptions
+                                    responseHandler:self
+                                        callOptions:_callOptions];
+  [_call start];
+}
+
+- (void)cancel {
+  dispatch_async(_dispatchQueue, ^{
+    if (_call) {
+      [_call cancel];
+      _call = nil;
+    }
+    if (_handler) {
+      id<GRPCResponseHandler> handler = _handler;
+      dispatch_async(handler.dispatchQueue, ^{
+        [handler closedWithTrailingMetadata:nil
+                                      error:[NSError errorWithDomain:kGRPCErrorDomain
+                                                                code:GRPCErrorCodeCancelled
+                                                            userInfo:@{
+                                                              NSLocalizedDescriptionKey :
+                                                                  @"Canceled by app"
+                                                            }]];
+      });
+      _handler = nil;
+    }
+  });
+}
+
+- (void)writeWithMessage:(GPBMessage *)message {
+  if (![message isKindOfClass:[GPBMessage class]]) {
+    [NSException raise:NSInvalidArgumentException format:@"Data must be a valid protobuf type."];
+  }
+
+  dispatch_async(_dispatchQueue, ^{
+    if (_call) {
```

Comment:
No because this could be queued before finish but dispatched after finish. I'll just drop the op.

Reply:
I don't see how a write after a finish can ever be correct. This is all public API, right? Would you not expect all writes  to be queued to _dispatchQueue before finish is called?

Reply:
Server can finish the call and a sendMessage may already be in the queue at that time. Of course we can limit the assert to only the case where the client proactively closes the connection, but that should use another variable rather than `_call` (I'm not sure if it's better than just dropping the call).